### ISSUE TITLE
Change parameter value after trial

### ISSUE THREAD

**Issue by eafpres** [optuna/optuna#3210]
I have an optimization of an xgboost model. I use early stopping, and the initial iters is set by an optuna sampler. I would like to update the iters value of the trial after the trial so that the optimizer "sees" the actual iterations used vs just the max.  I cannot find a way to do that.

### Environment

OS Ubuntu 20.04
Optuna 2.10.0
Python 3.8

**Comment by fsmosca** [comment]
What do you mean by iters? Can you show a minimal example code of what you are trying to achieve?

**Comment by eafpres** [comment]
@fsmosca:  Here is my objective function:

```
# train xgboost model
#
def train_xgboost(trial,
                  train_X,
                  train_Y,
                  val_X,
                  val_Y,
				  early_stopping,
                  study):
#
  objective = 'reg:gamma'
  eval_metric = 'gamma-deviance'
#
  n_iters = int(trial.suggest_discrete_uniform('n_iters', 10, 1000, 10))
#
  model = xgb.XGBRegressor(objective = objective,
                           eval_metric = eval_metric)
#
  eval_set = [(val_X, val_Y)]
#
  model.fit(train_X,
            train_Y,
            eval_set = eval_set,
            early_stopping_rounds = early_stopping,
            verbose = 0)
#
  results = model.evals_result()['validation_0'][eval_metric]
  result = results[-1]
  trial.set_user_attr('act_iters', len(results))
#
  return result
```

So the maximum rounds of boosting are set by n_iters in the trial.suggest() method, but the actual rounds is also affected by early stopping.  The early_stopping parameter tells the method to stop training if early_stopping_rounds pass without improvement in the objective.  

So my issue is in the overall optimization, for example using TPE optimizer, it is using the maximum values not the actual values, so whatever effect the number of iterations has interacting with other hyperparameters is lost.

So I wanted to update the n_iters value to 'act_iters" which is determined from the actual stopping round.

I know I can do this another way--set the max to a large value, then use
n_iters = int(trial.suggest_discrete_uniform('n_iters', act_iters, act_iters, 1) which will force it to act_iters.  However, I actually want the maximum to be varied, so this loses some of the functionality I have now.  If I could change the value of n_iters before it is stored back to the experiments which are used for the optimization, that would be a bit closer to what I want.

Note the for minimum repro I removed most of the actual hyperparameters used in the overall optimization; there are several more than n_iters in the actual code.

**Comment by fsmosca** [comment]
> So my issue is in the overall optimization, for example using TPE optimizer, it is using the maximum values not the actual values, so whatever effect the number of iterations has interacting with other hyperparameters is lost.

Sorry but your issue in unclear to me.

Have a look at the example below tell me where is the problem.
```python
stop_round = 20
pct = 0.1  # stop_round is 10% of min_iter
min_iter = stop_round / pct = 20/0.1 = 200

n_iters = int(trial.suggest_discrete_uniform('n_iters', min_iter, 1000, 10))
model = xgb.XGBRegressor(n_estimators=n_iters, ...)
model.fit(early_stopping_rounds=stop_round, ...)
```

**Comment by eafpres** [comment]
@fsmosca Thanks for looking at this again.  Let me try to explain using your sample code:

```
n_iters = int(trial.suggest_discrete_uniform('n_iters', min_iter, 1000, 10)) # the n_iters used by the TPE optimizer will be this value
model = xgb.XGBRegressor(n_estimators=n_iters, ...)
model.fit(early_stopping_rounds=stop_round, ...) # the actual iters will be determined by early stopping
```

So suppose we have two paramters, n_iters and reg_alpha, then the code is:

```
n_iters = int(trial.suggest_discrete_uniform('n_iters', min_iter, 1000, 10))
reg_alpha = trial.suggest_uniform(0.1, 0.2)
model = xgb.XGBRegressor(n_estimators=n_iters, reg_alpha = reg_alpha...)
model.fit(early_stopping_rounds=stop_round, ...)
```

Let's say we run 5 trials (ignore warmup trials), and they go as follows (assume result is, say, accuracy):
trial n_iters act_iters reg_alpha result
1     350      73          0.1            0.3
2     500      35          0.2            0.5
3     750      29          0.15          0.65
4     140      110        0.11          0.2
5     635      49          0.27          0.7

So in the above, in the actual cases accuracy improves with less iterations but more regularization, but the TPE optimizer only "sees" n_iters, and "thinks" more iterations is better. Assuming there is a complex non-linear high-dimensional space we are optimizing, then this would confuse the optimizer.  One option, of course, would be to leave n_iters out of the optimization space.  But it is hard to know if that would be optimal. So I want the optimizer to "see" act_iters, not n_iters.

**Comment by fsmosca** [comment]
I would suggest not to bother with act_iters and n_iters. What is important is the accuracy (the objective). Just let the optimizer do its job of finding the best param values in order to get the best accuracy.

**Comment by eafpres** [comment]
OK, I'll play with some options, but I think we have confirmed that changing trial parameters after the fact is not an option.

**Comment by sharmuz** [comment]
Has anything changed wrt this issue?

I'm trying to do the same thing @eafpres mentioned, albeit with catboost:

```python
def objective_catboost(
    trial: Trial,
    X_dev: pd.DataFrame,
    y_dev: pd.Series,
    X_val: pd.DataFrame,
    y_val: pd.Series,
) -> float:
    search_space = {
        "depth": trial.suggest_int("depth", 1, 10),
        "learning_rate": trial.suggest_loguniform("learning_rate", 1e-5, 1),
        # ...other params...
    }

    model = CatBoostRegressor(
        **search_space,
        loss_function="RMSE",
        eval_metric="RMSE",
    )

    pruning_callback = CatBoostPruningCallback(trial, "RMSE")
    model.fit(
        X_dev,
        y_dev,
        eval_set=(X_val, y_val),
        early_stopping_rounds=100,
        callbacks=[pruning_callback],
    )
    pruning_callback.check_pruned()
    
    ##########
    # How do I get num_trees out along with params sampled from search_space?
    ##########
    num_trees = model.get_best_iteration()

    preds = model.predict(X_val)
    score = mean_squared_error(y_val, preds, squared=False)

    return score
```

**Comment by eafpres** [comment]
> Has anything changed wrt this issue?
> 
> I'm trying to do the same thing @eafpres mentioned, albeit with catboost:

What I ultimately did is set a large default iters, capture the actual iters after early stopping, then call:

```trial.suggest_int("n_iters", actual_iters, actual_iters)```

after the model is fit, which makes it part of the optimization space.

**Comment by nabenabe0928** [comment]
I would like you to check the comment below:

- https://github.com/optuna/optuna/issues/4618#issuecomment-1758821285

In principle, it's not a good practice to tune `n_estimators` or `n_iters` in tree Ensemble based methods because it is obvious that the performance usually enhances or stabilizes with more trees.
It is also mentioned by a Kaggle Grandmaster as described in the comment above.

If you still want to tune the parameter, you can still do it, but as the performance yielded by a smaller `n_iters` will be worse than the bigger one (because the performance is cumulative optimal), anyways you will get a bigger `n_iters` later.

**Comment by eafpres** [comment]
> In principle, it's not a good practice to tune `n_estimators` or `n_iters` in tree Ensemble based methods because it is obvious that the performance usually enhances or stabilizes with more trees.
> It is also mentioned by a Kaggle Grandmaster as described in the comment above.

Actually, it's not about tuning n_estimators; the early stopping prevents over-fitting.  However, as I said in the original post, if there is some other complex non-linear interaction between other hyperparameters and the actual n_estimators, then it would seem beneficial to allow the optimizer to be aware of it.  

Without early stopping, you have to establish a value for n_estimators ahead of time, and you don't really know if it is too few or too many. So likely directly tuning it would slow down the optimization.  That is where early stopping is a better strategy.

**Comment by nabenabe0928** [comment]
Please first check [the tutorial](https://www.youtube.com/watch?v=VC8Jc9_lNoY&t=1018s) first.
It's obvious that `n_iters` or `n_estimators` has an interaction with some other parameters especially with `learning_rate`.
But it's already well known that it's better to set a small learning rate with a big `n_iters`.

You usually don't tune the training budget, i.e., `n_iters` here along with other parameters because you already know that `n_iters` is better when it's larger.

As mentioned, early stopping is outside of optimization and it should not be touched by Optuna, which is why `n_iters` should not be suggested by Optima.
There are already many ways to do early stopping in a white-box manner, so please use one of them.
Plus, even if you wrongly use Optuna.for early stopping and let it suggest `n_iters`, you should not report how many true iterations you did, but report what `n_iters` you set for the training.
